src/renderers/WebGLRenderer.js

```diff
@@ -1337,67 +1333,62 @@ function WebGLRenderer( parameters ) {
 		const parameters = programCache.getParameters( material, lights.state, shadowsArray, scene, object );
 		const programCacheKey = programCache.getProgramCacheKey( parameters );
 
-		let program = materialProperties.program;
-		let programChange = true;
+		let programs = materialProperties.programs;
 
-		// always update environment and fog - changing these trigger an initMaterial call, but it's possible that the program doesn't change
+		// always update environment and fog - changing these trigger an getProgram call, but it's possible that the program doesn't change
 
 		materialProperties.environment = material.isMeshStandardMaterial ? scene.environment : null;
 		materialProperties.fog = scene.fog;
 		materialProperties.envMap = cubemaps.get( material.envMap || materialProperties.environment );
 
-		if ( program === undefined ) {
+		if ( programs === undefined ) {
 
 			// new material
+
 			material.addEventListener( 'dispose', onMaterialDispose );
 
-		} else if ( program.cacheKey !== programCacheKey ) {
+			programs = new Map();
+			materialProperties.programs = programs;
 
-			// changed glsl or parameters
-			releaseMaterialProgramReference( material );
+		}
 
-		} else if ( materialProperties.lightsStateVersion !== lightsStateVersion ) {
+		let program = programs.get( programCacheKey );
 
-			programChange = false;
+		if ( program !== undefined ) {
 
-		} else if ( parameters.shaderID !== undefined ) {
+			// early out if program and light state is identical
 
-			// same glsl and uniform list
-			return;
+			if ( materialProperties.currentProgram === program && materialProperties.lightsStateVersion === lightsStateVersion ) {
```

Comment:
It seems a bit messy that currentProgram is checked here even though it's managed outside this function. Would it be fine to just set currentProgram at the end of this function instead? That would make the early out work also in WebGLRenderer.compile().

Reply:
I think that should work!